core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala

```diff
@@ -154,6 +154,113 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
   import org.bitcoins.core.psbt.InputPSBTRecord._
   import org.bitcoins.core.psbt.PSBTInputKeyId._
 
+  def nextRole(txIn: TransactionInput): PSBTRole = {
+    if (isFinalized) {
+      PSBTRole.ExtractorPSBTRole
+    } else {
+      (nonWitnessOrUnknownUTXOOpt, witnessUTXOOpt) match {
+        case (None, None) =>
+          PSBTRole.UpdaterPSBTRole
+        case (Some(_), None) | (None, Some(_)) | (Some(_), Some(_)) =>
+          val finalizeT = finalize(txIn)
+          finalizeT match {
+            case Failure(_) =>
+              val dummySigner = Sign.dummySign(ECPublicKey.freshPublicKey)
+              Try(toUTXOSigningInfo(txIn, dummySigner)) match {
+                case Failure(_) =>
+                  PSBTRole.SignerPSBTRole
+                case Success(_) =>
+                  PSBTRole.FinalizerPSBTRole
+              }
+            case Success(_) =>
+              PSBTRole.FinalizerPSBTRole
+          }
+      }
+    }
+  }
+
+  def prevOutOpt(vout: Int): Option[TransactionOutput] = {
+    witnessUTXOOpt match {
+      case Some(witnessUTXO) =>
+        Some(witnessUTXO.witnessUTXO)
+      case None =>
+        nonWitnessOrUnknownUTXOOpt match {
+          case Some(tx) =>
+            Some(tx.transactionSpent.outputs(vout))
+          case None => None
+        }
+    }
+  }
+
+  private def missingSigsFromScript(
```

Comment:
`@tailrec` and a scaladoc

Reply:
Can't do `@tailrec` because of the conditional :(